src/dropout.jl

```diff
@@ -0,0 +1,156 @@
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims::Colon)
+    T = real(eltype(dst))
+    val = convert(T, 1/(1-p))
+    rand!(rng, dst)
+    ## This is what we want, but it hits a SIMD bug, solved by _fast_broadcast!
+    # dst .= (dst.>p) .* val .* src
+    _fast_broadcast!(dst, src) do q, x
+        ((real(q)>p) * val) * x
+    end
+    dst
+end
+
+# For other dims, we we do need to allocate something.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims)
+    T = real(eltype(dst))
+    tmp = similar(dst, T, ntuple(d -> d in dims ? size(src,d) : 1, ndims(src)))
+    rand!(rng, tmp)
+    val = convert(T, 1/(1-p))
+    ## One-pass strategy -- faster on GPU
+    dst .= ((tmp.>p) .* val) .* src
+    ## Two-pass strategy -- slightly faster on some CPUs?
+    # _fast_broadcast!(tmp) do q
+    #     (q>p) * val
+    # end
+    # dst .= tmp .* src
+end
+
+# The gradient needs to keep the random choices made, thus store at least a BitArray,
+# but the following way turns out to be faster & simpler:
+function ChainRulesCore.rrule(::typeof(dropout), rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(real(eltype(A)))
+    val = convert(T, 1/(1-p))
+    keep = if dims isa Colon
+        similar(A, T) 
+    else
+        similar(A, T, ntuple(d -> d in dims ? size(A,d) : 1, ndims(A)))
+    end
+    rand!(rng, keep)
+    Y = @. ((keep>p) * val) * A
+    function dropout_back(Δ)
+        dY = unthunk(Δ)
+        dA = @. ((keep>p) * val) * dY
+        (NoTangent(), NoTangent(), dA, NoTangent())
+    end
+    return Y, dropout_back
+end
+# Possibly TODO: another approach to the gradient would be to copy the RNG
+# and then re-generate the same mask, instead of storing it. This saves memory
+# and seems about as fast, but needs a method `copy(::CUDA.RNG)` and careful checking.
+# https://github.com/FluxML/NNlib.jl/pull/454#issuecomment-1369357402
+
+"""
+    _fast_broadcast!(f, x, y, z...)
+
+This does `x .= f.(x, y, z...)`, but works around
+an issue with broadcasting that prevents SIMD in such cases.
+Can be removed once https://github.com/JuliaLang/julia/issues/43153 is fixed.
+
+Not intended for general use. Does not check sizes!
+"""
+function _fast_broadcast!(f::F, x::Array, yz...) where {F<:Function}
+    bc = Broadcast.instantiate(Broadcast.broadcasted(f, x, yz...))
+    @simd ivdep for I in eachindex(bc)
+        @inbounds x[I] = bc[I]
+    end
+    return x
+end
+function _fast_broadcast!(f::F, x::AbstractArray, yz...) where {F<:Function}
+    # CUDA does not suffer from this bug
+    broadcast!(f, x, x, yz...)
+end
+
+
+"""
+    _rng_from_array(x)
+
+Return the random number generator most appropriate for `x`:
+`CUDA.default_rng()` for `CuArray`, else `Random.default_rng()`
+"""
+_rng_from_array(::AbstractArray) = Random.default_rng()
+
+@non_differentiable _rng_from_array(::Any)
```

Comment:
Any reason not to copy https://github.com/FluxML/Flux.jl/blob/ee78ce3cefb027228413f8edace3c0385139d786/src/utils.jl#L36-L49 wholesale (minus the CUDA overload)?

Reply:
Just simplifying a bit, I couldn't figure out why there were so many functions. Why make a different choice on 1.6
```
julia> using Random

julia> Random.default_rng()
MersenneTwister(0x9687b6121c4ccb062f473c9c3c8bccc6)

julia> Random.GLOBAL_RNG
Random._GLOBAL_RNG()

julia> VERSION
v"1.6.0"
```
compared to master:
```
julia> using Random

julia> Random.default_rng()
TaskLocalRNG()

julia> Random.GLOBAL_RNG
Random._GLOBAL_RNG()

julia> VERSION
v"1.10.0-DEV.204"
```

Reply:
I don't remember now, but based on https://github.com/FluxML/Flux.jl/pull/1849#issuecomment-1022687847 it might've been related to thread safety?

Reply:
Cthulhu tells me that `rand(...)` uses `default_rng()` on 1.6 as well and it returns a thread-local RNG, so maybe this was much ado about nothing. cc @darsnack if I've missed something though.